src/scheduler.cpp

```diff
@@ -196,6 +196,9 @@ void SingleThreadedSchedulerClient::AddToProcessQueue(std::function<void ()> fun
         m_callbacks_pending.emplace_back(std::move(func));
     }
     MaybeScheduleProcessQueue();
+    if (!m_pscheduler->AreThreadsServicingQueue()) {
```

Comment:
In commit "Handle no threads serving in MaybeScheduleProcessQueue" (201eebf13b2723d7fe821d7222600fa04fadc0b0)

If we can make this code:

```c++
if (m_pscheduler->AreThreadsServicingQueue()) [
    MaybeScheduleProcessQueue();
} else {
    EmptyQueue();
}
```

it would be more obvious what's going on here.

Otherwise if the unconditional `MaybeScheduleProcessQueue` call is needed it would be good have a comment saying what it's needed for.

Reply:
Should it be inside `EmptyQueue`

Reply:
> Should it be inside `EmptyQueue`

I wouldn't change emptyqueue because it's called other places which don't need this logic, and I can't see how the behavior would fit in there (how you could coherently describe it or change the emptyqueue name to reflect it)

Reply:
Changed. However if the scheduler thread is stopped after `AreThreadsServicingQueue` then the function won't be called right?

Reply:
Test wallet_multiwallet fails due to this code path, see my changes in EmptyQueue, you should schedule process queue when event queue is not empty, the logic is reversed there.

Reply:
> I wouldn't change emptyqueue because it's called other places

Where? It's called in `FlushBackgroundCallbacks` only in init.cpp i'm pretty sure it will not affect anything wrong.

Reply:
> > I wouldn't change emptyqueue because it's called other places
> 
> Where? It's called in `FlushBackgroundCallbacks` only in init.cpp i'm pretty sure it will not affect anything wrong.

Not saying it's wrong, saying I don't understand how it makes sense for what how the EmptyQueue function is named, or what it currently does, or how it's called other places.

Maybe your suggestion to change EmptyQueue to schedule tasks instead of running them on the current thread is good, but it doesn't make sense to me right now, and since no rationale was given and I can't read your mind, I'm saying I wouldn't change it here.

Reply:
> Test wallet_multiwallet fails due to this code path, see my changes in EmptyQueue, you should schedule process queue when event queue is not empty, the logic is reversed there.

wallet_multiwallet.py is failing on appveyor with "'unloadwallet' RPC took longer than 60.000000 seconds. Consider using larger timeout for calls that take longer to return. (-344)"

https://ci.appveyor.com/project/DrahtBot/bitcoin/builds/31405838

So it looks like there is a problem with the current implementation. But I don't understand the comment about EmptyQueue or logic being reversed

Reply:
It makes much more sense compare to current version
```
void SingleThreadedSchedulerClient::EmptyQueue() {
    if (!m_pscheduler->AreThreadsServicingQueue())
        return;
    auto pendingCallbacks = [this]() -> bool {
        LOCK(m_cs_callbacks_pending);
        return !m_callbacks_pending.empty();
    };
    bool should_continue = pendingCallbacks();
    while (should_continue) {
        ProcessQueue();
        should_continue = pendingCallbacks();
    }
}
```
1. You don't need to do stupid things, like here one,
`if (!m_pscheduler->AreThreadsServicingQueue())`
2. You don't need to check against is there pending callbacks it's done there, so if you don't have pending ones you don't need to enter eventually endless waiting.

Reply:
Iv'e been trying to reproduce appveyor error wihtout any luck.

Reply:
> 1. You don't need to do stupid things, like here one,
>    `if (!m_pscheduler->AreThreadsServicingQueue())`
> 2. You don't need to check against is there pending callbacks it's done there, so if you don't have pending ones you don't need to enter eventually endless waiting.

@bvbfan, I'm confused by what you are saying here. You seem to be saying you don't need two things, but your code sample is doing both those things. Are you saying that for some reason it is better to do (1) in `EmptyQueue` instead of `AddToProcessQueue`? If so, what is the reason you think this? To me it seems better to do (1) in `AddToProcessQueue` instead of  `EmptyQueue`, because ` `AddToProcessQueue` has to be aware of thread state while `EmptyQueue` does not.

I'm even more confused what you are saying about (2) because you're saying we don't need a check for pending callbacks but your code is actually adding an extra check (before the first loop iteration).

More importantly, I can't figure out what this suggestion has to do with wallet_multiwallet.py. If you can clarify, it would be much appreciated. I've also been trying to reproduce the failure there and look for possible causes and haven't been able to figure it out.

Reply:
>  I'm confused by what you are saying here. You seem to be saying you don't need two things, but your code sample is doing both those things. 

I mean you don't need to do there things outside else, i was not clear. These 2 things should be done here in `EmptyQueue`

> More importantly, I can't figure out what this suggestion has to do with wallet_multiwallet.py

```
if (m_pscheduler->AreThreadsServicingQueue()) [
    MaybeScheduleProcessQueue();
} else {
    EmptyQueue();
}
```
This code is wrong, you try to process / flush events but it does not do that. Let analyse current code
 ```
void SingleThreadedSchedulerClient::EmptyQueue() {
assert(!m_pscheduler->AreThreadsServicingQueue()); // <- assert (you can avoid that)
    bool should_continue = true;
    while (should_continue) {
        ProcessQueue(); // <--- here is the bigger mistake, it tries to process events before it's checked that has pending callbacks, you can enter this and wait to infinity (since app is shutting down) it should be checked for pendings before process queue
        LOCK(m_cs_callbacks_pending);
        should_continue = !m_callbacks_pending.empty();
    }
```

Reply:
> ```
>        ProcessQueue(); // <--- here is the bigger mistake, it tries to process events before it's checked that has pending callbacks, you can enter this and wait to infinity (since app is shutting down) it should be checked for pendings before process queue
> ```

Explain this more? `ProcessQueue` doesn't wait for callbacks if there aren't any, it just runs a  callbacks if there is one.

> Iv'e been trying to reproduce appveyor error wihtout any luck.

I finally reproduced the problem cutting down the test and running in a loop with --usecli. Looking at the stack trace I see `unloadwallet` RPC thread waiting for scheduler thread `MaybeCompactWalletDB` call, which is stuck in `ReleaseWallet` calling `SyncWithValidationInterfaceQueue`, which is stuck because `SyncWithValidationInterfaceQueue` will always hang if called from the scheduler thread. This is basically the same deadlock that was in fanquake's backtrace from https://github.com/bitcoin/bitcoin/issues/16307#issuecomment-597669986, which suggests that 2a42fc1bdb393f88d72ec4d5c93b43b2b86dcc91 and 2097e35a635cd58bfc21e8ab149a903d58e94bf2 aren't going to work as fixes for this problem   in their current form. Maybe more can be done to get them working but an alternate approach like 06d2b530e57b88c0542e2bd8db00ca0b034b589d ([branch](https://github.com/ryanofsky/bitcoin/commits/pr/notify-shared), [comment](https://github.com/bitcoin/bitcoin/issues/16307#issuecomment-597302227)) might be easier at this point